### DISCUSSION TITLE
Mechanical scripting – coupling ACP and Ansys Mechanical

### DISCUSSION THREAD

**Discussion by gabbo842** [ansys/pymechanical#921] [Q&A] [answered]
Hi everyone,

I'm trying to write a pyAnsys script to analyse a composite structure. I'm launching a Mechanical session in which I import the geometry then a set the mesh and the boundary conditions. Then I save the input file (file .dat). After saving and closing the Mechanical session, with pyACP I import the input.dat file and I define the composite material. After defining the composite material, is there a way to upload the model and run the simulation in Ansys Mechanical or the only way is to solve the ACP model in a MAPDL session? In pyAnsys guide I found how to import and solve the ACP model in MAPDL but I would prefer to run the analysis in Mechanical, if it is possible. Thanks.

**Comment by koubaa** [comment]
pinging @greschd @janvonrickenbach

**Comment by greschd** [comment]
Hi @gabbo842,

It's possible to follow the same workflow with PyACP + PyMechanical as in Workbench (ACP + Mechanical), but both the code and corresponding example are still work in progress - so might be a bit tricky to get working. 
It currently relies on internals of Mechanical, which means that either:
- the workflow works only on Windows, with "remote" Mechanical
- you need to compile yourself a component that enables running on Linux + embedded

The latest version of the PyMechanical + PyACP example can be found here: https://github.com/ansys/pyacp/tree/doc/pymechanical_example_noshim/examples/pymechanical (note the feature branch `doc/pymechanical_example_noshim`).

As a starting point, you could try running the `remote_workflow_noshim.py` script (Windows only).

The broad overview of this workflow is that:
- Mechanical creates an input file (mesh) for ACP with `Model.InternalObject.WriteHDF5TransferFile`
- ACP defines the lay-up, and creates its output files for Mechanical and post-processing with `export_shell_composite_definitions` and `export_materials`
- This code blocks imports the plies into Mechanical: https://github.com/ansys/pyacp/blob/5c308eadcb28d690afd76b02bbb7922de0052599/examples/pymechanical/remote_workflow_noshim.py#L75
- In the example, post-processing is done with PyDPF Composites, which is built upon the same compute engine as Composites post-processing in Mechanical

We are aiming to make this example + workflow easier to use in the near future; probably by providing some helper functions to execute these steps.
Longer-term, we aim to provide native APIs in (Py)Mechanical for this communication with ACP. The timeline on that is less clear, however.

If you are willing to use the current (half-baked) state of this example, we'd be happy to help you along, and would appreciate your feedback.

As an aside: are you interested in only the shell workflow, or also for solid models? On the PyACP side that will be supported for 2025R1, but yet more work is required for the PyMechanical example.

**Comment by ekostson** [comment, marked as answer]
@gabbo842 - think you asked this in the AIS learning forum and I replied there (https://innovationspace.ansys.com/forum/forums/topic/mechanical-scripting-coupling-acp-and-ansys-mechanical/) - was hoping it was all clear, but if not is some more guidance - I referred there to the commands to import the cdb file, just as done in the pyacp->pymaodl example you refer to.

In addition to the above workflow (previous post) which is good:

So a sample code could look like the below (embedded pymech.) - it uses the pyacp+pymapdl example, but instead of mapdl it uses pymech. (pyacp is the same below as in the example: https://acp.docs.pyansys.com/version/dev/examples/gallery_examples/001_basic_flat_plate.html#sphx-glr-examples-gallery-examples-001-basic-flat-plate-py):

```
# Import the standard library and third-party dependencies.
import pathlib
import tempfile
import os
import smtplib
import ansys.mechanical.core as mech

# %%
# Import the PyACP dependencies.
from ansys.acp.core import (
    ACPWorkflow,
    PlyType,
    example_helpers,
    get_composite_post_processing_files,
    get_directions_plotter,
    get_dpf_unit_system,
    launch_acp,
    material_property_sets,
    print_model,
)

# sphinx_gallery_thumbnail_number = 3


# %%
# Launch PyACP etc..
```
#Pymechanicalstartshere
```
src=str(workflow.get_local_cdb_file())
dest = r'C:\\Users\\.......\\Downloads\\b2\\myacpcdb.cdb'
os.rename(src, dest)

work_dir = r'C:\\Users\\.......\\Downloads\\b2\\'
path_to_242 = r'C:\Program Files\ANSYS Inc\v242'
app = mech.App(version=242) # starts a non-graphical Mechanical session within the python.exe
globals().update(mech.global_variables(app)) # update global variables to get access to the same Model, DataModel, etc variables as in the Mechanical scripting consoleâ€
# Add static analysis


# Import cdb

mimp=Model.GeometryImportGroup.AddModelImport()
mimp.ProcessValidBlockedCDBFile=False
mimp.ModelImportSourceFilePath=dest
mimp.Import()
# etc....(loads,BC,solve) same as the pymapdl example from here on
```
